### PR TITLE
add save_traces utility function

### PR DESCRIPTION
**Pull request by bzm3r**
## Summary
Add a `save_traces` utility function, which saves trace output to a file which can be opened by Chrome's Trace Viewer. Location of output is determined by environment variable `TRACE_OUTPUT`.

## Related Issues
Currently trying to figure out where this function should go in druid-shell: https://github.com/xi-editor/druid/pull/35

Wondering if perhaps it should be a xi_trace utility?

## Review Checklist
- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

### CODE REVIEW

File: rust/trace/src/chrome_trace_dump.rs [hunk 1]
```diff
@@ -82,6 +82,42 @@ where
...
+    let traces = if sorted_chronologically {
+        super::samples_cloned_sorted()
+    } else {
+        super::samples_cloned_unsorted()
+    };
+
+    let trace_output_path = match env::var("TRACE_OUTPUT") {
+        Ok(output_path) => output_path,
+        Err(_) => {
+            println!("Environment variable TRACE_OUTPUT not set, defaulting to ./target/trace_output.trace");
```
Comment (cmyr) [inline rust/trace/src/chrome_trace_dump.rs]: all of these `println` should probably be `eprintln`... don't think it's likely to be a problem in this case, but as xi uses stdout to communicate if this was used there it would cause a crash.

File: rust/trace/src/chrome_trace_dump.rs [hunk 2]
```diff
@@ -82,6 +82,42 @@ where
...
+            String::from("./target/trace_output.trace")
```
Comment (cmyr) [inline rust/trace/src/chrome_trace_dump.rs]: I would discourage a fixed path name, which means you'll always overwrite any existing files. One compromise would be using the PID of the current process?

Comment (bzm3r) [reply rust/trace/src/chrome_trace_dump.rs]: @cmyr or perhaps something that includes time + date?

Comment (cmyr) [reply rust/trace/src/chrome_trace_dump.rs]: That's totallly an option; the downside is that it drags in another dependency.

Comment (bzm3r) [reply rust/trace/src/chrome_trace_dump.rs]: @cmyr How about we let the calling application decide how they want to name their path? I have changed the function so that it now expects a path, and thus doesn't default to anything---thoughts?

Comment (cmyr) [reply rust/trace/src/chrome_trace_dump.rs]: I think that is the best approach.

File: rust/trace/src/chrome_trace_dump.rs [hunk 3]
```diff
@@ -82,6 +83,36 @@ where
...
+/// using the Chrome browser by visiting the URL `about:tracing`. If `sorted_chronologically` is
+/// true then sort output traces chronologically by each trace's time of creation.
+pub fn save_traces(output_path: &Path, sorted_chronologically: bool) -> IOResult<()> {
+    let traces = if sorted_chronologically {
+        super::samples_cloned_sorted()
+    } else {
+        super::samples_cloned_unsorted()
+    };
+
+    match File::create(&output_path) {
```
Comment (cmyr) [inline rust/trace/src/chrome_trace_dump.rs]: I would return the error directly (`File::create(&output_path)?`) and let the caller handle it.

File: rust/trace/src/chrome_trace_dump.rs [hunk 4]
```diff
@@ -82,6 +83,36 @@ where
...
+        Ok(f) => f,
+        Err(err) => {
+            eprintln!("Could not create trace output file at: {}.", &output_path);
+            return IOResult(err);
+        }
+    };
+
+    match serialize(&traces, &mut trace_file) {
```
Comment (cmyr) [inline rust/trace/src/chrome_trace_dump.rs]: ooh, I didn't realize that this has a different result type. Luckily we actually already have a custom Error type in this crate, so we should be able to just use that as our return type. In this case as well we can let the caller handle the error, and we don't need to do any printing here.

File: rust/trace/src/chrome_trace_dump.rs (hunk 3 above)
Comment (cmyr) [reply rust/trace/src/chrome_trace_dump.rs]: Additionally, we should at least think about what the correct behaviour is in the case where the path exists.

File: rust/trace/src/chrome_trace_dump.rs [hunk 5]
```diff
@@ -82,6 +83,36 @@ where
     serde_json::from_reader(input).map_err(Error::Json)
 }
 
+/// Save tracing data to to supplied path, using the Trace Viewer format. Trace file can be opened
+/// using the Chrome browser by visiting the URL `about:tracing`. If `sorted_chronologically` is
+/// true then sort output traces chronologically by each trace's time of creation.
+pub fn save_traces(output_path: &Path, sorted_chronologically: bool) -> IOResult<()> {
```
Comment (cmyr) [inline rust/trace/src/chrome_trace_dump.rs]: I think the correct type for `output_path` is `<P: AsRef<Path>>`. This has the advantage of working with `&Path`, `&PathBuf`, `PathBuf`, `String`, `str`, `OsStr`, etc.

File: rust/trace/src/chrome_trace_dump.rs (hunk 3 above)
Comment (bzm3r) [reply rust/trace/src/chrome_trace_dump.rs]: @cmyr I gave some thought to what correct behaviour should be in the case where the path (file) already exists:

- we could overwrite the file: this is dangerous, since an important file could be overwritten.
- we could rename the file name component of the path: in the end, in order to guarantee that we come up with a unique file name, we cannot rely on any methods that use randomness at some level to generate a file name (as there would always exist some possibility of failure). We must instead list every file name in the directory containing the file, come up with a file name (possibly through some random means), and ensure it is not in the already tabulated list of existing file.
- we could do nothing, and abort (return an error): this is robust, in the sense that `save_traces` will definitely not screw up in any important way, and it is up to the caller to ensure they are providing okay file names.

I have chosen to go with the last option, for now. Let me know what you think.

**Comment by bzm3r** [comment]
@cmyr I think this process has been enlightening, because the save function we are left with is so barebones, after deciding what to relegate to the the caller, and what should be done by xi-trace, that I even wonder if the function is worth adding. Also, I am not quite sure how to elegantly deal with what Error type to return.

File: rust/trace/src/chrome_trace_dump.rs [hunk 6]
```diff
@@ -82,6 +83,40 @@ where
...
+            if md.is_file() {
+                return Err(Error::Io(IOError::from(IOErrorKind::AlreadyExists)));
+            }
+        }
+        Err(err) => {
+            return Err(Error::Io(err));
+        }
+    }
+
+    let mut trace_file = match File::create(&output_path) {
```
Comment (cmyr) [inline rust/trace/src/chrome_trace_dump.rs]: We should be able to write all (most) of these as, `let thing = Failable::call()?;` and the errors will be correctly propagated up.

**Comment by cmyr** [comment]
Yea, finding the dividing line between a client and a library's responsibilities can be tricky. That said, small improvements to ergonomics are worth doing, and if this makes it easier to dump a trace to disk I think that makes sense.

In the interests of treating this as an exercise: one thing we might consider is more closely mimicking the api used elsewhere, where we have methods in an `impl` of the `Trace` type, and then we expose global functions that call these methods on the shared global instance?

So basically we could have `Trace::save(&self, path, sorted) -> Result<()>` and then we just expose that from a global function.

But that's just an idea.


In any case I think we can clean up this function a bit more now by using `?` to handle the error propagation, but otherwise I'm happy!

**Comment by bzm3r** [comment]
@cmyr Do have a look when you have time, I hope to have addressed your most recent review comments.

File: rust/trace/src/lib.rs [hunk 7]
```diff
@@ -819,6 +821,22 @@ impl Trace {
...
     }
+
+    pub fn save<P: AsRef<Path>>(
+        &self,
+        path: P,
+        sort: bool,
+    ) -> Result<(), chrome_trace_dump::Error> {
+        let traces = if sort { samples_cloned_sorted() } else { samples_cloned_unsorted() };
+
+        if fs::metadata(&path)?.is_file() {
```
Comment (cmyr) [inline rust/trace/src/lib.rs]: what if it's a directory?

In anycase I would just suggest `if path.exists() { .. }`.

Also, in these fns where we take `AsRef<Path>`, the first thing I normally do is `let path = path.into()` so that we are working concretely with a `&Path`, and can use `Path` fns.

File: rust/trace/src/lib.rs [hunk 8]
```diff
@@ -819,6 +821,22 @@ impl Trace {
         samples.sort_unstable();
         samples
     }
+
+    pub fn save<P: AsRef<Path>>(
```
Comment (cmyr) [inline rust/trace/src/lib.rs]: I'd make this `pub(crate)` since we don't expect it to be used externally. One important reaeson to avoid `pub` when not necessary is that it makes it a semvar breaking change if we want to modify the api in the future.

File: rust/trace/src/lib.rs [hunk 9]
```diff
@@ -819,6 +821,23 @@ impl Trace {
         samples.sort_unstable();
         samples
     }
+
+    pub(crate) fn save<P: AsRef<Path>>(
```
Comment (cmyr) [inline rust/trace/src/lib.rs]: I think this actually probably _is_ public API now? sorry 😬

Comment (bzm3r) [reply rust/trace/src/lib.rs]: no problem! is done :)
